### DISCUSSION TITLE
Cleaning session after every test: what's the purpose of multiple tests then?

### DISCUSSION THREAD

**Discussion by dakur** [cypress-io/cypress#21645] [Questions and Help]
Hi,

I have to admit, I'm a bit confused about the new handling of session – clearing it every time a test function is finished – and I need some direction..

I always understood the spec file as a user scenario/journey and test functions as a features that are tested on this journey. It divided the test into labeled group which prevent big mess. Let's say:

- user scenario (what is it, spec file): manager's first campaign
- test functions (what the user journey consist of, test function in spec file):
  - create campaign
  - fill campaign text
  - upload material
  - add comment to material
  - insert material to campaign
  - publish the campaign

Each test function started in the point where the previous ended (it's user *journey*) and continued.

Now as the session is cleared, the continuity of tests is broken and I would have to wrap all code in all functions into `cy.session()` (which doesn't seem to be fulfilling the intention of it) or squash all test functions into one which create the big mess again.

My question is: am I missing something in the purpose of test functions? In which case would you use more of them? How would you deal with it?

Thank you in advance.

**Comment by dakur** [comment]
Anyone?

**Reply by stokrattt** [reply]
I have similar tests and  I continionly ise cypres v9 because new cypress v10 is broken it or it was very long test or squash all test functions into one

**Comment by bahmutov** [comment]
By using independent tests that set up the data they need, you can avoid the situation when one of the tests fails (like "upload material") and the rest _cannot run_. I would also suggest reading the Best Practices guide, since it talks about the tests that rely on the state left by the previous tests https://on.cypress.io/best-practices

**Reply by dakur** [reply]
@bahmutov Thank you for your response, but maybe we don't understand each other. I'm not testing particular components without context, I'm testing that user is able to go from point A to point B with X intermediate steps and that on the way there are no unexpected obstacles. Why on earth should I want to delete the state? Real user's state is purged neither when he's browsing my application and e2e tests should simulate user as much as it can. Or am I wrong?

**Reply by kdp88** [reply]
@dakur i really thought i was the only one out there who believes in everything you just wrote!!! an e2e spec file is the user journey not some cookie cutter way to say hey create a session in the before and visit /home  visit /profile visit /user-list, no human on earth would actually 'visit' a url,  they are going to click the profile button and epxect to land on /profile, they are going to click the dropdown that brings them to a user list.. say it with me.... An e2e spec is the user journey, its not bad practice to utilize it blocks in said spec to 'rely' on each other

**Comment by jozso39** [comment]
@dakur If you want to keep the session between tests, play with the `testIsolation` feature. You have to specify it on the top describe of the test:
```
describe(`your test`, { testIsolation: false }, () => {
		it(`this test or step`, () => {
			// test code
		});
	});
	
```

This way, the session will not be cleared between it-s.
There is another problem: the sessions will carry over to another tests. You just have to clear the session in the before hook.

**Reply by krishanUnacademy** [reply]
Thankyou so much i was stuck in this from long time..!
